### DISCUSSION TITLE
Framer Motion works on static routes and fails on dynamic routes

### DISCUSSION THREAD

**Discussion by Liam-hi** [vercel/next.js#46534] [Help] [answered]
Hi! I created two static pages (home & about) and a few dynamic routes. My goal is to animate the navigation bar whenever I navigate between the pages. For some reason, Framer Motion works well when I click on the static pages but when I click on the dynamic pages - i get following error: TypeError: retrieveData is undefined.

The typeerror (and back animation) disappears when I remove AnimatePresence from _app.js. Anyone have an idea what the problem is?

app.js 

```
export default function App({ Component, pageProps, router }) {
  return (
  <AnimatePresence mode='wait'><Component {...pageProps} key={router.pathname}/></AnimatePresence>
  );
}
```

component/navbar.js

export default function Navbar() {

```
  const router = useRouter();
  return (
    <div className="navigation-bar">
        <motion.p key="nav" exit={{transform: x: 10}} initial={{opacity: 1}}animate={{ y: 20}}  href="/">Home</Link> / <Link  href="/about">About</Link></motion.p>
    </div>
  )
}
```

slug.js 

```
export default function Details() {
    const router = useRouter(); 
    const retrieveData = data.find(x => x.id == router.query.id);

  return (
    <div>
      <Navbar />
      <h1>Retrieve data</h1>
      <h1>{retrieveData.name}</h1>
      
    </div>
  )
}
```

index.js 

```
export default function Home() {

  const populate = data.map(x =>
    <Link key={x.id} href={`/${x.id}`}><li>
      {x.name}
    </li></Link>
  );

  return (
    <>
       <Navbar />
       { populate }
    </>
  )
}
```

**Comment by ciruz** [comment]
What is ``data`` in ``slug.js`` and how/where does it come from?

To me it sounds like ``data`` is not defined at all, therefore find will not be able to locate anything, and when you try to access ``retrieveData.name`` this is not defined as well.

**Comment by Liam-hi** [comment]
Hi ciruz, I didn't write it in order to make the code more readable. Sometimes router.query.id is undefined and.I think that's causing the problem. 

utils/data.js

```
const data = 
[
    { name: 'Nima', id: 1 },
    { name: 'Li', id: 2 },
    { name: 'Lisa', id: 3 },
];

export default data;
```

**Comment by icyJoseph** [comment]
Hi,

Could you try to confirm that the issue is not the first frame having `{}`?

> query: Object - The query string parsed to an object, including [dynamic route](https://nextjs.org/docs/routing/dynamic-routes) parameters. It will be an empty object during prerendering if the page doesn't use [Server-side Rendering](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props). Defaults to {}

For example:

```jsx
export default function Details() {
    const router = useRouter(); 
    const retrieveData = data.find(x => x.id == router.query.id);

  if (!retrieveData) return null

  return (
    <div>
      <Navbar />
      <h1>Retrieve data</h1>
      <h1>{retrieveData.name}</h1>
      
    </div>
  )
}
```

If the problem is because of the query `{}` -> `{ id: 'foo' }` step, then we might be able to act accordingly, but let's confirm first.

**Comment by Liam-hi** [comment]
Hi @icyJoseph, 

thank you for your reply. it doesn't seem like it but the crashing stopped . The exit animation won't work though (can't identify why?).

So whenever I click on a link (the dynamic ones), the animation and the exit animation works perfectly. When I click on a new link, or the back button in the browser, the application crashes.

Here's the live code, I'm not sure if it's going to help me or not. 

https://codesandbox.io/p/sandbox/vigorous-drake-n1go2f?file=%2Fpages%2F_app.tsx&selection=%5B%7B%22endColumn%22%3A1%2C%22endLineNumber%22%3A14%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A14%7D%5D

**Comment by icyJoseph** [comment, marked as answer]
Ugh, I fear this has to do with shared mutable state, but I refuse to believe... 

I think since, Framer's AnimatePresence:

> `AnimatePresence` enables the animation of components that have been removed from the tree.

When the route changes, the `useRouter` hooks returns the value for the next path, and your Details component, reads from the router, the new value. 

- https://github.com/framer/motion/issues/1375

It looks like it started to happen to other people too. Some claiming the `onExitComplete` callback is never triggering.

This has nothing to do with dynamic routes.

In your About page place this log:

```js
console.log("About:", useRouter().asPath);
```

So that it looks like:

```jsx
export default function About() {
  console.log("About:", useRouter().asPath);
  return (
    <div>
      <Navbar />
      <h1>About page or whatever</h1>
    </div>
  );
}
```

Now navigate, you'll see `About: /` on the logs. Suspicious.

I don't want to suggest "hacks" just yet, but this is not looking good.

**Reply by icyJoseph** [reply]
I couldn't resist a hack, please try with:

```jsx
import Navbar from "../components/Navbar";
import data from "../utils/data.js";
import { useRouter } from "next/router";
import { useRef } from "react";
import { useIsPresent } from "framer-motion";

export default function Details() {
  const router = useRouter();
  const ref = useRef();
  const isPresent = useIsPresent();
  ref.current = isPresent ? router.query.id : ref.current;
  const retrieveData = data.find((x) => x.id == ref.current);

  return (
    <div>
      <Navbar />
      <a className="explore">
        <span
          onClick={(e) => {
            router.back();
          }}
        >
          Back
        </span>
      </a>
      <h1>Retrieve data</h1>
      <h2>{retrieveData?.name}</h2>
    </div>
  );
}
```

**Reply by icyJoseph** [reply]
@Liam-hi Ah so it still fails... :/  -

It is sort of bad news, and perhaps "shared mutable state" is too broad. What happens is that AnimatePresence keeps a reference to the now unmounted component, but then `useRouter` updates and it triggers a re-render of the component. Maybe the useEffect is not needed, and it can be done with `isPresence` directly. I also forgot the `router.query.id` as dependency, ugh, 🤦 hacks are funky.

**Reply by icyJoseph** [reply]
@Liam-hi  so the failure you see now is because of `router.query` being `{}` initially 😅 let's not forget that!

I've updated the hack example, removed the effect, and did a `retrieveData?.name` on the last heading.

Summary of hack changes:

```js
  const router = useRouter();
  const ref = useRef();
  const isPresent = useIsPresent();
  ref.current = isPresent ? router.query.id : ref.current;
  const retrieveData = data.find((x) => x.id == ref.current);

```

```js
  <h2>{retrieveData?.name}</h2>
```

**Comment by Liam-hi** [comment]
Thank you for dedication @icyJoseph and for helping me identify the problem. Is the shared mutable state-thing bad news? 
Interesting, I wasn't aware of this issue. Thank you for the link. The hack works like a charm but when I refresh the page, I get a type error: TypeError: Cannot read properties of undefined (reading 'name')

@icyJoseph Ah, I get it! Brilliant and creative solution by the way. Thank you^1000 🙏
